engine/src/dot/http_rpc.rs

```diff
@@ -70,6 +72,46 @@ impl RpcClientT for PolkadotHttpClient {
 	}
 }
 
+/// Adds a default port to the url based on the scheme (http, https, ws, wss),
+/// if none exists. Otherwise preservers existing port.
+///
+/// This function assumes that `url` is already validated, i.e.:
+///  - It's accepted by `Url::parse()`
+///  - It includes a host part
+///  - It does not have a fragment part
+fn ensure_port(url: SecretUrl) -> Result<SecretUrl> {
+	// we use url::Url to get the default port for our scheme
+	let targetport = Url::parse(url.as_ref())
+		.unwrap() // We know that `url` is valid
+		.port_or_known_default()
+		.ok_or(anyhow::anyhow!("Unknown scheme and no port given."))?;
+
+	// We use http::uri::Uri to ensure that the default port is added if none exists
+	// We split the uri into components, insert the port, and reconstruct.
+	let mut parts = url
+		.as_ref()
+		.parse::<Uri>()
+		.unwrap() // We know that `url` is valid
+		.into_parts();
+
+	// Update the authority part of the uri by mapping over it.
+	let authority = parts.authority.unwrap();
+	parts.authority =
+		if authority.port().is_none() {
+			Some(format!("{}:{}", authority.as_str(), targetport).parse().map_err(|err| {
+				anyhow!("Unexpected error when trying to append port to url: {err}")
+			})?)
+		} else {
+			Some(authority)
+		};
```

Comment:
can dedup the Some()
```suggestion
	parts.authority = Some(if authority.port().is_none() {
		format!("{}:{}", authority.as_str(), targetport)
			.parse()
			.map_err(|err| anyhow!("Unexpected error when trying to append port to url: {err}"))?
	} else {
		authority
	});
```

Reply:
Rewrote execution flow.